Add the ability to growfs from the filesystem module optionally.

Firstyear opened the pull request:
Add the flag "growfs" to the filesystem module, allowing a filesystem to be grown. This works really nicely with lvol, to create a work flow such as:
```
- name: Create lvol
  sudo: yes
  action: lvol lv=test vg=vg size=1G state=present

-name: lvol filesystem
  sudo: yes
  action: filesystem fstype=xfs dev=/dev/mapper/vg-test growfs=yes
```
Changing the lvol size, automatically resizes the FS from, and when you initially create the lvol, creates it too.

gregdek commented:
@abulimov please review for functionality and adherence to Ansible module guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If you give +1, we will merge the changes.

abulimov commented on system/filesystem.py:
```diff
@@ -41,6 +41,12 @@
+    choices: [ "yes", "no" ]
+    default: "no"
+    description:
+    - If yes, if the block device and filessytem size differ, grow the filesystem into the space. This DOES NOT support shrinking. Will only grow if the FS is mounted.
```
The FS should be mounted only in the case of XFS filesystem. We could check if fstype=XFS and volume is mounted in this module, and just resize FS in other cases. Also, while XFS cannot be shrinked, we can rename growfs to resizefs and support shrinking via resize2fs. In this case, we should check that FS is not mounted, and run `e2fsck -f /dev/some_disk` before shrinking. Option to shrink FS makes sense, because `lvol` module supports shrinking.

Firstyear replied:
The comment about mounted FS should really be a warning about XFS specifically. It doesn't actually check if it's mounted or not, so the warning would make more sense.

I don't think you can express shrinking in a playbook. Consider this playbook structure.

Create VG
Create Lvol
Create Filesystem with growfs

So, now you change the Lvol size, and it grows. All good. 

So what if you reduce the lvol size? You would reduce the block size BEFORE you shrink the FS. So you would actually need to re-organise to:

Create VG
Create Filesystem with growfs
Create Lvol

Now it will shrink. But at the same time you have removed the ability to grow unless you ran the play twice ... 

I think that shrinking will probably need more consideration, and would likely need direct integration with the lvol module.

abulimov replied:
Ansible usage is not limited to playbooks. It is often used as orchestration engine, for massive actions on remote hosts. For me, shrinking makes sense in orchestration.

As for comment about mounted FS, we could change it to mention XFS, or (better) add check to module, so running ansible in check mode could warn us about not mounted XFS.

Firstyear replied:
True, it is. I'm just thinking about the general case. It seems like a lot of extra work and potential states to deal with in a way, but I can see in the long term it's the right solution. I'll work on shrink as well then, but for now, only ext*.

Yeah, I'll add the check then for xfs, and word the warning differently.

abulimov replied:
Thanks, it would be really great!

abulimov commented:
Current version is good, but it would be better to improve functionality before merging. Or at least fix description about mounted FS.

gregdek commented:
Looks like at the very least the description needs to be changed, so marking this "needs revision" for now. How much of a revision is up to @Firstyear I guess. :)

Firstyear commented:
Okay, I've done another push. I've fixed some of the doc comments about XFS, changed some command mapping bits to clear up a few things and to provide extensibility later.

Right now, resizing for shrink is just too hard.
You need to potentially unmount, fsck, resize. There are so many more edge cases, potential issues and complexities. Right now, btrfs is the easiest, with an online shrink capability.

So I've made it able to support shrink in the future, but I think that it is just a really hard thing to achieve at this time without some better thought and time.

alxgu commented on system/filesystem.py:
```diff
@@ -63,17 +66,68 @@ def main():
         supports_check_mode=True,
     )
 
+    # There is no "single command" to manipulate filesystems, so we map them all out and their options
```
You can use fsadm for fs-check and resize.
It supports ext2, ext3, ext4, ReiserFS and XFS and in near feature btrfs.

fsadm -y resize /dev/vg/test

gregdek commented:
@abulimov can you re-review?

abulimov commented on system/filesystem.py:
```diff
@@ -82,9 +136,18 @@ def main():
+    elif fs == fstype and resizefs == True:
+        cmd = module.get_bin_path(growcmd, required=True)
+        if module.check_mode:
+            changed = True
```
shouldn't we just exit now with changed=True? Now, if we are in check mode, execution will continue till next  check for module.check_mode on line 156. This flow seems not very obvious and error-prone in future.

abulimov commented on system/filesystem.py:
```diff
@@ -41,9 +41,12 @@
     - If yes, allows to create new filesystem on devices that already has filesystem.
     required: false
-  opts:
+  resizefs:
```
It seems that you have accidentally removed "opts" options from docs

abulimov commented on system/filesystem.py:
```diff
@@ -82,9 +136,18 @@ def main():
+        if module.check_mode:
+            changed = True
+        else:
+            rc,out,err = module.run_command("%s %s" % (cmd, dev))
```
shouldn't we check rc?

gregdek commented:
Back to you, @Firstyear.

Firstyear commented:
Hi,

Fixes for your comments have been provided.

abulimov commented on system/filesystem.py:
```diff
@@ -41,6 +41,12 @@
+    choices: [ "yes", "no" ]
+    default: "no"
+    description:
+    - If yes, if the block device and filessytem size differ, grow the filesystem into the space. Note: XFS Will only grow if mounted.
```
The build has failed because of this line. As you use ': ' symbol in description string, it becomes dict after YAML parsing in the process of generating documentation. You can check it in any yaml parses, for example [here](https://yaml-online-parser.appspot.com/). To fix this you can just put description line in quotes, or remove ': ' symbol.

abulimov commented:
Except one small issue with documentation YAML, everything looks good to me, thanks!

Firstyear commented:
Changed from : to ,

gregdek commented:
Hmmm... fails travis?

abulimov commented:
Strange, but Travis now fails on rendering docs for bundler plugin, which is untouched by this request.

gregdek commented:
Yep, looking at it.

gregdek commented:
So it's likely to be breaking because of an unquoted colon in the doc strings; our yaml parser doesn't like that. We just introduced the doc string parser into Travis, so we're probably picking up bleedover from other docstrings that are broken now. We'll get back to this soon...

gregdek commented:
Closed and reopened to kick Travis.

gregdek commented:
Kicking Travis

gregdek commented:
@Firstyear so looks like you got caught in a bug in documentation elsewhere; we committed a new Travis check, and your code contains old code in bundler that's breaking the commit. Could you please rebase? Sorry for the confusion.

Firstyear commented:
I had some issues with this rebase. Please review and make sure the pull request only changes filesystem.py.

Firstyear commented:
Yeah, it's accidentally caught my ovirt changes because I did those on devel, and I needed to rebase onto devel. Any ideas on how to remove that from the pull request?

gregdek commented:
Hmm... cache the changes to ovirt, blow it away, rebase again?

Firstyear commented:
Hi,

Git just confuses and infuriates me, so it was easier to just make a new clean branch, and re-submit the pull req. See https://github.com/ansible/ansible-modules-extras/pull/672
